bin/inx-chronicle/src/api/v1/routes.rs

```diff
@@ -0,0 +1,636 @@
+// Copyright 2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::str::FromStr;
+
+use axum::{
+    extract::{Extension, Path},
+    routing::*,
+    Router,
+};
+use chronicle::{
+    bson::U64,
+    chrysalis::{
+        payload::transaction::TransactionPayload,
+        prelude::{OutputId, TransactionId},
+    },
+    db::model::inclusion_state::LedgerInclusionState,
+    BsonExt, DocExt,
+};
+use futures::TryStreamExt;
+use mongodb::{
+    bson::{doc, DateTime, Document},
+    options::FindOptions,
+    Database,
+};
+use time::OffsetDateTime;
+
+use super::{
+    extractors::{MessagesQuery, OutputsQuery},
+    responses::{Record, TransactionResponse, Transfer, *},
+};
+use crate::api::{
+    error::APIError,
+    extractors::{Expanded, Pagination, TimeRange},
+    APIResult,
+};
+
+pub fn routes() -> Router {
+    Router::new()
+        .nest(
+            "/messages",
+            Router::new()
+                .route("/", get(messages_query))
+                .route("/:message_id", get(message))
+                .route("/:message_id/raw", get(message_raw))
+                .route("/:message_id/metadata", get(message_metadata))
+                .route("/:message_id/children", get(message_children)),
+        )
+        .nest(
+            "/outputs",
+            Router::new()
+                .route("/", get(outputs_query))
+                .route("/:transaction_id/:idx", get(output)),
+        )
+        .nest(
+            "/transactions",
+            Router::new()
+                .route("/:message_id", get(transaction_for_message))
+                .route("/history/:address", get(transaction_history))
+                .route("/included-message/:transaction_id", get(transaction_included_message)),
+        )
+        .route("/milestones/:index", get(milestone))
+        .nest("/analytics", Router::new().route("/addresses", get(address_analytics)))
+}
+
+async fn message(database: Extension<Database>, Path(message_id): Path<String>) -> APIResult<MessageResponse> {
+    let mut rec = database
+        .collection::<Document>("chrysalis_messages")
+        .find_one(doc! {"message_id": &message_id.to_string()}, None)
+        .await?
+        .ok_or(APIError::NoResults)?;
+    let mut message = rec.take_document("message")?;
+    Ok(MessageResponse {
+        network_id: message.get_as_u64("network_id")?,
+        parents: message
+            .take_array("parents")?
+            .iter()
+            .map(|m| m.as_string())
+            .collect::<Result<_, _>>()?,
+        payload: message.take_bson("payload").ok().map(Into::into),
+        nonce: message.convert_document::<U64, _>("nonce")?.into(),
+    })
+}
+
+async fn message_raw(database: Extension<Database>, Path(message_id): Path<String>) -> APIResult<Vec<u8>> {
+    let mut rec = database
+        .collection::<Document>("chrysalis_messages")
```

Comment:
Should we put some `&str`s into constants? Especially those that are used several times?

Reply:
yes